### PR TITLE
LPS-121855

### PR DESCRIPTION
**Pull request by jonmak08**
Hey @liferay-frontend,

Attached is an update for http://issues.liferay.com/browse/LPS-121855.

**Test Plan:**

1. Login to Liferay
2. Click on the **Application Menu**
3. Click on **Accounts**
4. Add an account
5. Click on the **Application Menu**
6. Click on **Account Users**
7. Click on the `+` to select an account user
8. Click on **Choose**
9. Create a new user

admin-users-account

@markocikos, @julien, @ambrinchaudhary,

I made a fair bit of changes to my original pull request since I forgot to migrate 2 other sub folders.  I made each sub folder as separate commits just to break things up otherwise there would be a significant amount of changes under 1 commit.  Most of the changes are pretty straight forward, but just figured I would get another pair of eyes on this since there are quite a few file changes in this one. 

Please let me know if you have any questions. Thanks!

### CODE REVIEW

**Comment by jonmak08** [comment]
Hey @kresimir-coko,

This is the pull request that will have merge conflicts with the changes that you are pushing in with https://github.com/liferay-frontend/liferay-portal/pull/746.  If everything goes well in here, I am ready to push this up tomorrow for final review and testing.  Let me know if I should hold off for your pull request or if you want me to do anything in particular with your changes.

Please let me know if you have any questions. Thanks!

**Review by markocikos — APPROVED** [review]
LGTM!

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_users_admin/js/AccountUsersAdminManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_users_admin/js/AccountUsersAdminManagementToolbarPropsTransformer.js:29]: Just like @markocikos pointed in https://github.com/liferay-lima/liferay-portal/pull/1507#discussion_r577635010

> See https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571
> 
> If it's just one line in if block, postForm, it's hard to justify an early return.

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: I know this was done inline and maybe that is why we should keep it this way, but just wondering.... since there is the `managementToolbarDisplayContext` attribute, the idea behind that was to have everything inside the java class and not to add other props to the tag, so, shouldn't we have a `getAdditionalProps` method in the `ManagementToolbarTag`, that could be overriden in each DisplayContext? I guess the same applies to the `propsTransformer`... Otherwise, we are having something "hibrid", with all the context in the DisplayContext but some extras outside it... 

What do you guys think about it, @markocikos @julien? Maybe we should ask to some backend developer...

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/js/AccountGroupAccountEntriesManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/js/AccountGroupAccountEntriesManagementToolbarPropsTransformer.js:44]: Same like before, see https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/js/AccountGroupAccountEntriesManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/js/AccountGroupAccountEntriesManagementToolbarPropsTransformer.js:70]: Same, https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/js/AccountGroupsManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/js/AccountGroupsManagementToolbarPropsTransformer.js:37]: Same, see https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountEntriesManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountEntriesManagementToolbarPropsTransformer.js:20]: Same like before, see https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountEntryAddressesManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountEntryAddressesManagementToolbarPropsTransformer.js:37]: Same, see https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountOrganizationsManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountOrganizationsManagementToolbarPropsTransformer.js:37]: Same here, see https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountRolesManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountRolesManagementToolbarPropsTransformer.js:37]: See https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountUsersManagementToolbarPropsTransformer.js
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/js/AccountUsersManagementToolbarPropsTransformer.js:45]: Same, https://github.com/liferay-frontend/liferay-portal/pull/759#discussion_r574501571

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp
Comment (ambrinchaudhary) [inline modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp:50]: I guess this should be inside the displayContent, there is a `isShowCreationMenu` method...

**Review by ambrinchaudhary — COMMENTED** [review]
@jonmak08 Just minor comments about the `if` conditions. 

But about the comments related to the DisplayContext, what should be inside or not, I think we should discuss it separately... don't know what you think about it

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp
Comment (julien) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: @ambrinchaudhary In my opinion concerning this task, the less Java code we have to touch the better: the goal is to get the management toolbar migrated to the new version. If we want to change our scope and do the migration plus some refactoring I doubt we'll finish this in a reasonable amount of time.

Comment (ambrinchaudhary) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: Well, it is not really about touching more or less java code, is about to create a consistent API and I think the `ManagementToolbarTag` was part of this task, isn't it? 
Because right now, we are kind of breaking the "API", as for me the `managementToolbarDisplayContext` implies that everything must be inside it...

Comment (jbalsas) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: As usual, finding the right balance in low-level components is key. One thing we're trying to do is to make it easier to navigate between the JS/Java realms.

It makes sense that `additionalProps` can be passed from the `DisplayContext` as well, but we might want to leave the decision to use them to the different teams. Some teams feel more comfortable being heavy on the backend side and would prefer to move it there while others might prefer to keep it in the JSP. 

Our job is to give them the tools to do so in a way they're productive while also making sure that complexity doesn't explode and that we have thousands of different ways to do the same thing.

Comment (ambrinchaudhary) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: I see @jbalsas, but me point is: there are two ways to pass data to the taglib: 
1) by adding each different attribute to the taglib, like `actionDropdownItems`, `additionalProps`, `filterDropdownItems`, `showInfoButton`,... 
2) or by adding `managementToolbarDisplayContext`, which should imply that everything is inside de displayContext. 

Mixing them (which is the case) is not to have a good API, but something hybrid wich may confuse developers... If you all are ok with this... then it's fine ;-)

Comment (julien) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: > Well, it is not really about touching more or less java code

Well if you have to edit Java code, I understand the opposite.

> Because right now, we are kind of breaking the "API", as for me the managementToolbarDisplayContext implies that everything must be inside it...

As far as I understand it, breaking the API would be removing or deprecating methods and attributes. 
In this case we have an API (that accepts works with attributes because that's how tags work, correct me if I'm wrong) and one of those attributes is called `managementToolbarDisplayContext` which can be used in conjunction with the other attributes. So I don't understand what are we breaking here?
Another thing about the `managementToolbarDisplayContext` is that we can't declare the JavaScript callbacks and so at least we'll have to use `managementToolbarDisplayContext` and `propsTransformer`.

I think that this `managementToolbarDisplayContext` attribute was added to make it easier for  developers to instantiate the tag via Java code, but if we have to move all the variable declared in the JSP to make it work or complete the migration it doesn't seem like an ideal solution for anyone. 

Of course we can always decide that starting from now we should put everything in the `managementToolbarDisplayContext` but as I said above some things won't be possible and it also wouldn't make sense to have other attributes that we don't use.

Comment (ambrinchaudhary) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: > `managementToolbarDisplayContext` which can be used in conjunction with the other attributes.

This is the only part I disagree...

Comment (julien) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: No problem, we can keep on talking about it until we find a solution everyone is happy with.

I wanted to say that when I wrote my comment, I hadn't seen @jbalsas 's [comment](https://github.com/liferay-frontend/liferay-portal/pull/807#discussion_r578343522), which sums up what I also think about finding the right balance and providing the correct tools.

If we can only use `managementToolbarDisplayContext` and no other attributes how are we going to deal with JS callbacks currently handled by `propsTransformer`?

(If there's a solution, I'm all for it, but it doesn't seem like it's going to be possible.)

Comment (jbalsas) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: > > managementToolbarDisplayContext which can be used in conjunction with the other attributes.
>
> This is the only part I disagree...

Hey @ambrinchaudhary, this one has a bit more context that is probably not here. Although it is a bit of a stretch, it can be linked to [[Discussion] High-level vs low-level component API](https://github.com/liferay/clay/issues/2275).

You can think of `managementToolbarDisplayContext` as a High Level API and the individual props as a Low Level API. Higher levels usually help providing more complex setups and defaults while lower levels allow for better composability and reusability.

As infrastructure, one of our biggest pain points in the past has been to provide very specific and closed APIs. I'm sure you've experienced them when trying to extend specific components (ie: management toolbar filters or additional bars). Thus, time ago we decided to try to work in a more layered way so that teams could have escape hatches when a very closed/high level component wouldn't suit them.

This is a similar case other than it deals with a declarative API and not with visual components composition. As infrastructure we often want to provide a reasonable set of options. If something is egregiously bad, we will definitely work to prevent its usage. I don't think this is such a case. 

Both a _DisplayContext_ and a _Tag Attribute_ are valid and commonly accepted patterns. As frontend infrastructure we are not in a position to force every team to structure their code so they can only use one single object. We leave it up to them because they might have valid reasons to pick one over the other that we might not know of and preventing it doesn't give us an obvious benefit over allowing it.

> If we can only use managementToolbarDisplayContext and no other attributes how are we going to deal with JS callbacks currently handled by propsTransformer?

@julien, this should have nothing to do with it. All attributes are being passed to different parts of the component. It makes no difference if it happens through different attributes or through a shared big bag of properties such as `DisplayContext`. In the end, `propsTransformer` is a string which could be also passed from `DisplayContext.getPropsTransformer()` and `additionalProps` is a list of key-value pairs that could be passed as `DisplayContext.getAdditionalProps()`.

What I think it's important here is to look at this from an API design perspective and the context in which they are expected to be used. You usually want to group together things that are related. For better or for worse, `DisplayContext` might have long abandoned that category. We could just rename it as `AllKindOfPropertiesBag` and drop in there anything new that we can think of. Tags are an API and expose attributes. Developers know how to use them and expect some specific ergonomics out of them. We want to look at our components in context. Necessarily forcing a specific Java object to be created is a far more restricting API than what a JSP tag consumer might expect (even though we might grow used to it here in Liferay).

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp
Comment (markocikos) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp:50]: @ambrinchaudhary  It may be that the product team forgot to move this to `DisplayContext`, but this also may be intentional. For example, `selectAccountUsersManagementToolbarDisplayContext` may be used by multiple management toolbars. If that is the case, this would be a way of overriding the default.

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp
Comment (jbalsas) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: Adding some additional context into how `propsTransformer` and `additionalProps` were born for those brave enough to go through the 200+ comments in [[LPS-112460, LPS-112460, LPS-116360] Updates DropdownMenu and DropdownActions tags and adds AdditionalProps mechanism to clay:* ta](https://github.com/liferay-frontend/liferay-portal/pull/2)

Comment (julien) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: > @julien, this should have nothing to do with it ...

True, I got confused, sorry about that

Comment (ambrinchaudhary) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: > Both a DisplayContext and a Tag Attribute are valid and commonly accepted patterns

I completely agree. That is why I said there are two options to send the data to the tag, and it is completely fine. My point was, that in my understanding the purpose of `managementToolbarDisplayContext` attribute was to not have other tag attributes with him, if you see in most cases this attribute is "alone", like in https://github.com/ambrinchaudhary/liferay-portal/blame/master/modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_scheduled.jsp#L34-L34. This other example, for instance, is a bad use of it: https://github.com/liferay/liferay-portal/blame/master/modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs_admin/view_images.jsp#L59-L59, because there are attributes duplicated like `actionDropdownItems`, or `filterDropdownItems` in the tag and in the displayContext, which makes no sense and could create confusion. 

Just to say, before writting this I spoke to @4lejandrito that was the one (with Drew I think) who worked in defining the purpose of this `managementToolbarDisplayContext`. 

But, as I said before, if you are all agree and happy about "mixing" this two approachs for sending the data through a tag, I have nothing more to add.

Comment (jbalsas) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_groups_admin/account_group/view_account_entries.jsp:53]: > My point was, that in my understanding the purpose of managementToolbarDisplayContext attribute was to not have other tag attributes with him, if you see in most cases this attribute is "alone", like in [...]
>
> Just to say, before writting this I spoke to @4lejandrito that was the one (with Drew I think) who worked in defining the purpose of this managementToolbarDisplayContext.

Well... that was the initial idea, but that would've been for the `clay:management-toolbar-wrapper` class as you can see in [LPS-81988 Create a wrapper taglib that just accepts a display context to wrap the management toolbar tag](https://github.com/brianchandotcom/liferay-portal/pull/59760). The params `clay:management-toolbar` would've stayed.

This predates the time where we were more thorough with our public conversations in the team, so there's no public record of my conversations with Drew, but from my own Slack:

> Drew Brokke
> no worries
> so you would prioritize the other attributes above the DC?
> 
> Chema Balsas
> yeah, because the dc has defaults, right?
> if I ser a totalItems value in the tag
> that’s more specific
> so we should use the dc if the value is null in the map
> 
> Drew Brokke
> but why not just override the totalItems value in the DC?
> instead of passing an DC and another attribute
> 
> Chema Balsas
> you could
> but if you wanted
> you might have a common dc where only a value changes
> and you can avoid additional hierarchies
> extending ad infinitum is not a good pattern either

This is the design that eventually landed through [LPS-81988 SF](https://github.com/brianchandotcom/liferay-portal/pull/59825) (I'm simply adding the link here to shame my past self).

> But, as I said before, if you are all agree and happy about "mixing" this two approachs for sending the data through a tag, I have nothing more to add.

Just to reiterate, we should be neither happy nor unhappy with it unless there's something patently wrong with it. Being more of a stylistic preference, from infra we probably want to avoid making a choice that's hard to undo and might force teams to bend over backwards to make something simple work they way the expect it.

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_users_admin/js/AccountUsersAdminManagementToolbarPropsTransformer.js
Comment (jonmak08) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_users_admin/js/AccountUsersAdminManagementToolbarPropsTransformer.js:29]: Sounds good. I will make that change.

I read that comment differently than you as if it was a 1 line or just a few lines of code, we would wrap the `if` check around the entire code block, but if it was more than that, we would return early.  I prefer wrapping the entire code block since that was our practice before, but I agree with the early return justification.

Thanks for clarifying this!

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp
Comment (jonmak08) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp:50]: I only added that there because in the file previously, the toolbar handler was wrapped in an `if` check on whether to be included or not.  See https://github.com/liferay-frontend/liferay-portal/pull/807/files/dd79ff48f9feddfb1704a083527b10adf8352bab..2dfcfea9d82435381e6ab7d871e5123eac8e62e8#diff-eaf417c5b06d208194a5c763e75fa438627504d8a480a75851ec63dd2bc61506L135.

I figured since we are migrating everything and the toolbar always needs to show, I would try to migrate their existing logic over, but if you feel it is better to have this in the `DisplayContext` and utilize the `isShowCreationMenu` method I can move it there.

**Review by kresimir-coko — APPROVED** [review]
Didn't test the changes locally, but went through the code and it all seems 🍑 to me.

File: modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp
Comment (ambrinchaudhary) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp:50]: If this was like this before, I'm ok, sorry for the confusion, I didn't see how was it previously

Comment (jonmak08) [reply modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_entries_admin/select_account_users.jsp:50]: No problem.  Just wanted to make sure you understood the reasoning behind the change.  I am open to your suggestions on how to make things better, so these conversations are very good to me to think twice about everything we are doing and find ways to improve.

**Comment by jonmak08** [comment]
New Pull Request: https://github.com/drewbrokke/liferay-portal/pull/1312

**Pull request closed without merging**
